### ISSUE TITLE
HttpClient: An asterisk for wildcards in NO_PROXY is not handled properly

### ISSUE THREAD

**Issue by boldt** [dotnet/runtime#61840]
### Description

When an asterisk in `NO_PROXY` is used, the `HttpClient` is not handling it properly. E.g., when `NO_PROXY=*.mydomain.com` is set, the `HttpClient` still connects to the proxy instead of the server.

I observed this issue with grpc-dotnet, which uses the `HttpClient`. In my case, the `NO_PROXY` is ignored and a connectiin to the proxy happens instead of the server directly. Thus, the HTTP/2 connection fails. They told me, to raise the issue here:

* https://github.com/grpc/grpc-dotnet/issues/1440

### Reproduction Steps

Assume the following MWE: 

```C#
static void Main(string[] args)
{
    try
    {
        var client = new HttpClient();
        var response = client.GetAsync(args[0]);
        var code = response.Result.StatusCode;
        Console.WriteLine(code);
    }
    catch (Exception ex)
    {
        Console.WriteLine(ex.Message);
    }
}
```

Run the MWE with an asterisk in the `NO_PROXY`:

```
HTTP_PROXY=http://proxy NO_PROXY=*.mydomain.com ./Example.exe http://example.mydomain.com
// or
HTTPS_PROXY=http://proxy NO_PROXY=*.mydomain.com ./Example.exe https://example.mydomain.com
```

### Expected behavior

Looking into Wireshark, I should see that the `NO_PROXY` is handled properly, e.g., the `HttpClient` connects to the server directly.

### Actual behavior

Looking into Wireshark, I can see that the `NO_PROXY` is ignored. 

* for HTTP: The `HttpClient` sends the HTTP-request to the proxy instead of direct connection the server
* for HTTPS: The `HttpClient` sends a CONNECT-request to the proxy instead of direct TLS connection the server

### Regression?

n/a

### Known Workarounds

A workaround is adding the full domain of the destination server, which is really dirty, since the env-var is getting longer for each service, sinde the wildcard does not work:

```
HTTP_PROXY=http://proxy NO_PROXY=example.mydomain.com ./Example.exe http://example.mydomain.com
// or
HTTPS_PROXY=http://proxy NO_PROXY=example.mydomain.com ./Example.exe https://example.mydomain.com
```

Like this, the `HttpClient` connects to the server directly. Thus, the `NO_PROXY` seems to be handled.

### Configuration

* .NET version: 5.0.403
* OS: Windows 10
* Architecture: x64

**Comment by wfurt** [comment]
I think it is incorrect expectation. .NET implementation mimics `curl` (as curl used to power our implementation on Linux) and `curl` does not support either. 

```
$ export NO_PROXY=*.microsoft.com
$ export ALL_PROXY=127.0.0.1:8080
$ curl -v http://www.microsoft.com/
*   Trying 127.0.0.1...
* TCP_NODELAY set
* connect to 127.0.0.1 port 8080 failed: Connection refused
* Failed to connect to 127.0.0.1 port 8080: Connection refused
* Closing connection 0
curl: (7) Failed to connect to 127.0.0.1 port 8080: Connection refused
```
If you want wildcard (I assume so) just skip the asterisk.
https://curl.se/libcurl/c/CURLOPT_NOPROXY.html

```
proxy-test$ export NO_PROXY=.microsoft.com
proxy-test$ ~/dotnet-latest/dotnet run http://www.microsoft.com/
result=StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Accept-Ranges: bytes
  ETag: "6082151bd56ea922e1357f5896a90d0a:1425454794"
  Server: AkamaiNetStorage
  X-Akamai-Transformed: 9 - 0 pmb=mTOE,1
  Cache-Control: max-age=0, no-cache, no-store
  Pragma: no-cache
  Date: Fri, 19 Nov 2021 22:20:04 GMT
  Connection: keep-alive
  Set-Cookie: _abck=2F8F6B1FBEF53D1DF1E3A029819CC4D4~-1~YAAQEZA6F63kISR9AQAAYQlJOgZZmyKA17smhW9DA7oY+T2ddjAbaD00Yfl5yiTQ41yu36YH1DkcsLS/JcsumWnM5aXxhhDLDm0cTrh7qOhORNiXVQVrAyzDLk04tufRTYkWJL8XJRPOGCKpUtkw3tsWudNSBrwp367us4E8jDJ6jzfWJy6OkQ8F1t23oloMRjHa3lDY3i7udrby4b9b0EHp1RvC05fd/HUlpBwcdGb9YJpOHQWoODoz2CHhf4WBpKCb7quggy2AaOUPnISF1LRcxDxRmYdHuPyykSxoeoYV9zfmb9shuPMofQv/zZLyYBY1kBm1d9aeTmCUS1CSTEQFFZZcFcI=~-1~-1~-1; Domain=.microsoft.com; Path=/; Expires=Sat, 19 Nov 2022 22:20:04 GMT; Max-Age=31536000
  Set-Cookie: bm_sz=7E4BBBF38612F9C666AF5B7C11C6D698~YAAQEZA6F67kISR9AQAAYQlJOg1DbVD0Mp/+w1OscCH5vsGmabSSzCaFVmpRjgAE5fycKPUrf3CeuDo4OV82qbLqeQ6KXE50QYyDC0Ul8zJDH1839a6w5ADFX7nw6C6u27RhRQCBmZJ7DGPmJFv3BtVNZ+ymY15eWMaO+4nqc6CCuSZhBIvt3LSa91sSTxtYOlPy1QdYxt1x39qF9idAUjT66IARptLwQkuNO/E6MnZGDbmh/NWSGEaJgwqzHQUQ3ZwzKBwgf2lvl2AVEhH79TQFMI7Zf8AZ9r4kkeksSx9gohIEwiI=~3421252~4474422; Domain=.microsoft.com; Path=/; Expires=Sat, 20 Nov 2021 02:20:04 GMT; Max-Age=14400
  Content-Type: text/html
  Last-Modified: Wed, 04 Mar 2015 07:39:54 GMT
  Expires: Fri, 19 Nov 2021 22:20:04 GMT
  Content-Length: 1120
}
```

let me know if that works for you need @boldt
cc @JamesNK

**Comment by boldt** [comment]
Hey @wfurt,

thanks for your response. You're 100% correct, I want wildcards.

I made a test an I can confirm, that the leading `.` in the `no_proxy` works as expected:

```
HTTP_PROXY=http://proxy NO_PROXY=.mydomain.com ./Example.exe http://example.mydomain.com
// or
HTTPS_PROXY=http://proxy NO_PROXY=.mydomain.com ./Example.exe https://example.mydomain.com
```

I also found the blog post [We need to talk: Can we standardize NO_PROXY?](https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/) from the colleagues at gitlab, which also described the issue, that the `NO_PROXY` is handled differently by differnet tools/languages.

**Comment by JamesNK** [comment]
Is there documention for using HTTP_PROXY + HttpClient?

**Comment by boldt** [comment]
As a side note: In the gitlab blog they recommend the following for `no_proxy`:

> However, if you ever have to manage a stack written in multiple languages, you might need to consider setting HTTP proxy configurations to the lowest common denominator:
> 
> 1. Use lowercase form.
> 2. Use comma-separated hostname:port values.
> 3. IP addresses are okay, but hostnames are never resolved.
> 4. Suffixes are always matched (e.g. example.com will match test.example.com).
> 5. If top-level domains need to be matched, avoid using a leading dot (.).
> 6. Avoid using CIDR matching since only Go and Ruby support that.

Especially point 5: *avoid using a leading dot (.)*

**Comment by wfurt** [comment]
I was thinking about it more it it would be trivial to strip leading `*` to make it just work. 
And of course improved documentation with examples would help.
I'll keep this open for triage.

**Comment by boldt** [comment]
👍 for stripping the leading * and for an improved documentation

**Comment by karelz** [comment]
Triage: We should not add support for start and deviate from curl / Linux ecosystem pattern.
We should document it - by linking curl/Linux docs + giving couple of examples.

**Comment by vjrantal** [comment]
In our testing, we found the following difference compared to `curl` behavior.

When using `curl` to fetch `http://www.microsoft.com` a value without leading dot will match, such as `NO_PROXY=microsoft.com`. When using the same value with `dotnet`, there is no match.

The test script output [here](https://gist.github.com/vjrantal/6a46292705085bbafe7f0985eb5d4fb4#file-output-txt) can be seen the difference when a proxy value without leading dot is used.

If the intent is to exactly match the `curl` behavior, I was wondering should a new issue be created to track aligning this detail as well?

This also relates to the [earlier comment](https://github.com/dotnet/runtime/issues/61840#issuecomment-975371824) where a recommendation was given to avoid a leading dot. If one would do so, they would observe a behavior difference between `curl` and `dotnet`.

**Comment by antonfirsov** [comment]
I opened https://github.com/dotnet/dotnet-api-docs/pull/8363 and https://github.com/dotnet/docs/pull/31029.

After reading the thread, I actually think it would make sense to follow up on https://github.com/dotnet/runtime/issues/61840#issuecomment-1011903215, and consider fully matching the curl behavior. This is reflected in the wording of my doc change proposal.

/cc @wfurt

**Comment by antonfirsov** [comment]
The documentation PR-s are merged, I'm closing this in favor of #75199.
